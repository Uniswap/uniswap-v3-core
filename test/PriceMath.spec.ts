import {Contract, BigNumber} from 'ethers'
import {waffle} from '@nomiclabs/buidler'

import {expect} from './shared/expect'
import snapshotGasCost from './shared/snapshotGasCost'
import {encodePrice, expandTo18Decimals} from './shared/utilities'

import PriceMathTest from '../build/PriceMathTest.json'

describe.only('PriceMath', () => {
  const [wallet] = waffle.provider.getWallets()
  const deployContract = waffle.deployContract

  let priceMath: Contract
  before(async () => {
    priceMath = await deployContract(wallet, PriceMathTest, [])
  })

  describe('#getInputToRatio', () => {
    describe('edge cases', () => {
      it('0 all', async () => {
        await expect(priceMath.getInputToRatio(0, 0, 0, [0], true)).to.be.revertedWith('FixedPoint: DIV_BY_ZERO')
        await expect(priceMath.getInputToRatio(0, 0, 0, [0], false)).to.be.revertedWith('FixedPoint: DIV_BY_ZERO')
      })

      it('returns 0 if wrong direction', async () => {
        const price = encodePrice(expandTo18Decimals(75), expandTo18Decimals(1))

        // going from zero to one
        // that means reserve1 will decrease and reserve0 will increase
        // i.e. the price will decrease, so the target price must be lower than the current price
        const [amountIn] = await priceMath.getInputToRatio(
          expandTo18Decimals(1),
          expandTo18Decimals(50),
          30,
          [price],
          true
        )
        expect(amountIn).to.eq(0)
      })

      it('returns 0 if price is equal', async () => {
        const price = encodePrice(expandTo18Decimals(50), expandTo18Decimals(1))

        const [amountIn] = await priceMath.getInputToRatio(
          expandTo18Decimals(1),
          expandTo18Decimals(50),
          30,
          [price],
          false
        )
        expect(amountIn).to.eq(0)
      })

      it('gas: returns 0 if price is equal', async () => {
        const price = encodePrice(expandTo18Decimals(50), expandTo18Decimals(1))

        await snapshotGasCost(
          priceMath.getGasCostOfGetInputToRatio(expandTo18Decimals(1), expandTo18Decimals(50), 3000, [price], false)
        )
      })
    })

    describe('invariants', () => {
      for (const {priceTarget, reserve0, reserve1, lpFee, zeroForOne, summary} of [
        {
          priceTarget: encodePrice(expandTo18Decimals(50), expandTo18Decimals(1)),
          reserve0: BigNumber.from(1000),
          reserve1: BigNumber.from(100000),
          lpFee: 60,
          zeroForOne: true,
          summary: '1:100 to 1:50 at 60bps with small reserves',
        },
        {
          priceTarget: encodePrice(expandTo18Decimals(50), expandTo18Decimals(1)),
          reserve0: expandTo18Decimals(1),
          reserve1: expandTo18Decimals(100),
          lpFee: 60,
          zeroForOne: true,
          summary: '1:100 to 1:50 at 60bps',
        },
        {
          priceTarget: encodePrice(expandTo18Decimals(75), expandTo18Decimals(1)),
          reserve0: expandTo18Decimals(1),
          reserve1: expandTo18Decimals(100),
          lpFee: 45,
          zeroForOne: true,
          summary: '1:100 to 1:75 at 45bps',
        },
        {
          priceTarget: encodePrice(expandTo18Decimals(50), expandTo18Decimals(1)),
          reserve0: expandTo18Decimals(1),
          reserve1: expandTo18Decimals(100),
          lpFee: 30,
          zeroForOne: true,
          summary: '1:100 to 1:50 at 30bps',
        },
        {
          priceTarget: encodePrice(expandTo18Decimals(100), expandTo18Decimals(1)),
          reserve0: expandTo18Decimals(1),
          reserve1: expandTo18Decimals(50),
          lpFee: 200,
          zeroForOne: false,
          summary: '1:50 to 1:100 at 200bps',
        },
        {
          priceTarget: encodePrice(expandTo18Decimals(75), expandTo18Decimals(1)),
          reserve0: expandTo18Decimals(1),
          reserve1: expandTo18Decimals(50),
          lpFee: 60,
          zeroForOne: false,
          summary: '1:50 to 1:75 at 60bps',
        },
        {
          priceTarget: BigNumber.from('5192296858534827628530496329220096'),
          reserve0: BigNumber.from(101),
          reserve1: BigNumber.from(101),
          lpFee: 32,
          zeroForOne: false,
          summary: 'minimum tokens in both to tick 32 price, no fee',
        },
      ]) {
        describe(summary, () => {
          let priceBeforeSwap: BigNumber
          let amountIn: BigNumber
          let amountInLessFee: BigNumber
          let amountOut: BigNumber
          let amountOutMax: BigNumber
          let priceAfterSwap: BigNumber
          let priceAfterSwapWith1MoreWeiEffectiveInput: BigNumber

          before('compute swap result', async () => {
            priceBeforeSwap = encodePrice(reserve1, reserve0)
            ;[amountIn, amountOutMax] = await priceMath.getInputToRatio(
              reserve0,
              reserve1,
              lpFee,
              [priceTarget],
              zeroForOne
            )
            amountInLessFee = amountIn.mul(BigNumber.from(10000).sub(lpFee)).div(10000)
            amountOut = await (zeroForOne
              ? priceMath.getAmountOut(reserve0, reserve1, amountInLessFee)
              : priceMath.getAmountOut(reserve1, reserve0, amountInLessFee))

            // cap the output amount, if necessary
            if (amountOut.gt(amountOutMax)) amountOut = amountOutMax

            priceAfterSwap = zeroForOne
              ? encodePrice(reserve1.sub(amountOut), reserve0.add(amountInLessFee))
              : encodePrice(reserve1.add(amountInLessFee), reserve0.sub(amountOut))

            const outputFor1MoreWeiInput = await (zeroForOne
              ? priceMath.getAmountOut(reserve0.add(amountInLessFee), reserve1.sub(amountOut), 1)
              : priceMath.getAmountOut(reserve1.add(amountInLessFee), reserve0.sub(amountOut), 1))

            priceAfterSwapWith1MoreWeiEffectiveInput = zeroForOne
              ? encodePrice(reserve1.sub(amountOut).sub(outputFor1MoreWeiInput), reserve0.add(amountInLessFee).add(1))
              : encodePrice(reserve1.add(amountInLessFee).add(1), reserve0.sub(amountOut).sub(outputFor1MoreWeiInput))
          })

          it('snapshot', () => {
            // for debugging, store all the calculations
            expect({
              priceTarget: priceTarget.toString(),
              reserve0: reserve0.toString(),
              reserve1: reserve1.toString(),
              lpFee,
              zeroForOne,
              priceBeforeSwap: priceBeforeSwap.toString(),
              amountIn: amountIn.toString(),
              amountInLessFee: amountInLessFee.toString(),
              amountOut: amountOut.toString(),
              priceAfterSwap: priceAfterSwap.toString(),
              priceAfterSwapWith1MoreWeiEffectiveInput: priceAfterSwapWith1MoreWeiEffectiveInput.toString(),
            }).to.matchSnapshot()
          })

          it('price moves in the right direction', () => {
            if (zeroForOne) {
              expect(priceAfterSwap).to.be.lt(priceBeforeSwap)
            } else {
              expect(priceAfterSwap).to.be.gt(priceBeforeSwap)
            }
          })

          it('price after swap does not pass price target', () => {
            if (zeroForOne) {
              expect(priceAfterSwap).to.be.gte(priceTarget)
            } else {
              expect(priceAfterSwap).to.be.lte(priceTarget)
            }
          })

          it('1 more wei of effective input exceeds the next price', async () => {
            if (zeroForOne) {
              expect(priceAfterSwapWith1MoreWeiEffectiveInput).to.be.lt(priceTarget)
            } else {
              expect(priceAfterSwapWith1MoreWeiEffectiveInput).to.be.gt(priceTarget)
            }
          })

          it('gas', async () => {
            await snapshotGasCost(
              priceMath.getGasCostOfGetInputToRatio(reserve0, reserve1, lpFee, [priceTarget], zeroForOne)
            )
          })
        })
      }
    })
  })
})
