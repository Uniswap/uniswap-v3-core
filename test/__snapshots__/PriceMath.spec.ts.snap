// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceMath #getInputToRatio edge cases gas: returns 0 if price is equal 1`] = `517`;

exports[`PriceMath #getInputToRatio invariants 1:50 to 1:75 at 60bps gas 1`] = `9660`;

exports[`PriceMath #getInputToRatio invariants 1:50 to 1:75 at 60bps snapshot 1`] = `
Object {
  "amountIn": "11305074013661420926",
  "amountInLessFee": "11237243569579452400",
  "amountInPlus1": "11305074013661420927",
  "amountInPlus1LessFee": "11237243569579452401",
  "amountOut": "183503419072273966",
  "amountOutWith1MoreWeiEffectiveInput": "183503419072273966",
  "lpFee": 60,
  "priceAfterSwap": "389422264390112071185901382456528806",
  "priceAfterSwapWith1MoreWeiEffectiveInput": "389422264390112071192260621404761995",
  "priceBeforeSwap": "259614842926741381426524816461004800",
  "priceTarget": "389422264390112072139787224691507200",
  "reserve0": "1000000000000000000",
  "reserve1": "50000000000000000000",
  "zeroForOne": false,
}
`;

exports[`PriceMath #getInputToRatio invariants 1:50 to 1:100 at 200bps gas 1`] = `9670`;

exports[`PriceMath #getInputToRatio invariants 1:50 to 1:100 at 200bps snapshot 1`] = `
Object {
  "amountIn": "21133345019035461633",
  "amountInLessFee": "20710678118654752400",
  "amountInPlus1": "21133345019035461634",
  "amountInPlus1LessFee": "20710678118654752401",
  "amountOut": "292893218813452475",
  "amountOutWith1MoreWeiEffectiveInput": "292893218813452475",
  "lpFee": 200,
  "priceAfterSwap": "519229685853482762118747969201292707",
  "priceAfterSwapWith1MoreWeiEffectiveInput": "519229685853482762126090985838499876",
  "priceBeforeSwap": "259614842926741381426524816461004800",
  "priceTarget": "519229685853482762853049632922009600",
  "reserve0": "1000000000000000000",
  "reserve1": "50000000000000000000",
  "zeroForOne": false,
}
`;

exports[`PriceMath #getInputToRatio invariants 1:100 to 1:50 at 30bps gas 1`] = `9771`;

exports[`PriceMath #getInputToRatio invariants 1:100 to 1:50 at 30bps snapshot 1`] = `
Object {
  "amountIn": "415459942199694131",
  "amountInLessFee": "414213562373095048",
  "amountInPlus1": "415459942199694132",
  "amountInPlus1LessFee": "414213562373095049",
  "amountOut": "29289321881345247519",
  "amountOutWith1MoreWeiEffectiveInput": "29289321881345247569",
  "lpFee": 30,
  "priceAfterSwap": "259614842926741381723916990267895141",
  "priceAfterSwapWith1MoreWeiEffectiveInput": "259614842926741381356766158407536695",
  "priceBeforeSwap": "519229685853482762853049632922009600",
  "priceTarget": "259614842926741381426524816461004800",
  "reserve0": "1000000000000000000",
  "reserve1": "100000000000000000000",
  "zeroForOne": true,
}
`;

exports[`PriceMath #getInputToRatio invariants 1:100 to 1:50 at 60bps gas 1`] = `9771`;

exports[`PriceMath #getInputToRatio invariants 1:100 to 1:50 at 60bps snapshot 1`] = `
Object {
  "amountIn": "416713845445769666",
  "amountInLessFee": "414213562373095048",
  "amountInPlus1": "416713845445769668",
  "amountInPlus1LessFee": "414213562373095049",
  "amountOut": "29289321881345247519",
  "amountOutWith1MoreWeiEffectiveInput": "29289321881345247569",
  "lpFee": 60,
  "priceAfterSwap": "259614842926741381723916990267895141",
  "priceAfterSwapWith1MoreWeiEffectiveInput": "259614842926741381356766158407536695",
  "priceBeforeSwap": "519229685853482762853049632922009600",
  "priceTarget": "259614842926741381426524816461004800",
  "reserve0": "1000000000000000000",
  "reserve1": "100000000000000000000",
  "zeroForOne": true,
}
`;

exports[`PriceMath #getInputToRatio invariants 1:100 to 1:50 at 60bps with small reserves gas 1`] = `4103`;

exports[`PriceMath #getInputToRatio invariants 1:100 to 1:50 at 60bps with small reserves snapshot 1`] = `
Object {
  "amountIn": "417",
  "amountInLessFee": "414",
  "amountInPlus1": "418",
  "amountInPlus1LessFee": "415",
  "amountOut": "29278",
  "amountOutWith1MoreWeiEffectiveInput": "29328",
  "lpFee": 60,
  "priceAfterSwap": "259695628309264554133616521495830006",
  "priceAfterSwapWith1MoreWeiEffectiveInput": "259328624442666670080217128324129063",
  "priceBeforeSwap": "519229685853482762853049632922009600",
  "priceTarget": "259614842926741381426524816461004800",
  "reserve0": "1000",
  "reserve1": "100000",
  "zeroForOne": true,
}
`;

exports[`PriceMath #getInputToRatio invariants 1:100 to 1:75 at 45bps gas 1`] = `9771`;

exports[`PriceMath #getInputToRatio invariants 1:100 to 1:75 at 45bps snapshot 1`] = `
Object {
  "amountIn": "155399837648670547",
  "amountInLessFee": "154700538379251529",
  "amountInPlus1": "155399837648670548",
  "amountInPlus1LessFee": "154700538379251530",
  "amountOut": "13397459621556135322",
  "amountOutWith1MoreWeiEffectiveInput": "13397459621556135397",
  "lpFee": 45,
  "priceAfterSwap": "389422264390112072153277207641951089",
  "priceAfterSwapWith1MoreWeiEffectiveInput": "389422264390112071478778060119756628",
  "priceBeforeSwap": "519229685853482762853049632922009600",
  "priceTarget": "389422264390112072139787224691507200",
  "reserve0": "1000000000000000000",
  "reserve1": "100000000000000000000",
  "zeroForOne": true,
}
`;
